File: backend/src/main/java/org/schemeguard/backend/exception/ApiError.java
package org.schemeguard.backend.exception;

import java.time.OffsetDateTime;
import java.util.Map;

public record ApiError(
        int status,
        String message,
                OffsetDateTime timestamp,
                String errorCode,
                Map<String, String> fieldErrors
) {
        public ApiError(int status, String message, OffsetDateTime timestamp) {
                this(status, message, timestamp, null, null);
        }
}

File: frontend/src/shared/api/ApiError.ts
export type FieldErrors = Record<string, string | string[]>;

export class ApiError extends Error {
    readonly status: number | null;
    readonly errorCode?: string;
    readonly fieldErrors?: FieldErrors;

    constructor(
        message: string,
        status: number | null = null,
        errorCode?: string,
        fieldErrors?: FieldErrors
    ) {
        super(message);
        this.name = "ApiError";
        this.status = status;
        this.errorCode = errorCode;
        this.fieldErrors = fieldErrors;
    }
}